### PR TITLE
REUSE flag functionality

### PR DESCRIPTION
**Pull request by julianbermudez**
According with the XRCE standard, entities creation support to flag `REPLACE` and `REUSE`.
For the `REUSE` flag, it is needed to compare previous entities with the new entity.
It was achieved adding new `matched` functions which compare the attributes of the entities using the operator `equal to` added to [Fast RTPS](https://github.com/eProsima/Fast-RTPS/pull/273).
On the other hand, new tests have been added to check this functionality.

### CODE REVIEW

File: src/cpp/datareader/DataReader.cpp [hunk 1]
```diff
@@ -301,6 +268,40 @@ void DataReader::onNewDataMessage(eprosima::fastrtps::Subscriber* /*sub*/)
...
+            {
+                rv = (new_attributes == old_attributes);
+            }
+            break;
+        }
+        default:
+            break;
+    }
+
+    return rv;
```
Comment (BorjaOuterelo) [inline src/cpp/datareader/DataReader.cpp]:
```suggestion
    return new_attributes == old_attributes;
```

File: src/cpp/datawriter/DataWriter.cpp [hunk 2]
```diff
@@ -200,5 +165,39 @@ void DataWriter::onPublicationMatched(fastrtps::Publisher*, fastrtps::rtps::Matc
...
+            {
+                rv = (new_attributes == old_attributes);
+            }
+            break;
+        }
+        default:
+            break;
+    }
+
+    return rv;
```
Comment (BorjaOuterelo) [inline src/cpp/datawriter/DataWriter.cpp]:
```suggestion
    return new_attributes == old_attributes;
```

File: src/cpp/participant/Participant.cpp [hunk 3]
```diff
@@ -97,5 +112,39 @@ void Participant::onParticipantDiscovery(eprosima::fastrtps::Participant*, epros
...
+            {
+                rv = (new_attributes == old_attributes);
+            }
+            break;
+        }
+        default:
+            break;
+    }
+
+    return rv;
```
Comment (BorjaOuterelo) [inline src/cpp/participant/Participant.cpp]:
```suggestion
    return new_attributes == old_attributes;
```

File: src/cpp/topic/Topic.cpp [hunk 4]
```diff
@@ -83,5 +92,42 @@ void Topic::release(ObjectContainer& root_objects)
...
+                     (generic_type_.m_isGetKeyDefined == (new_attributes.getTopicKind() ==
+                                                          fastrtps::rtps::TopicKind_t::WITH_KEY));
+            }
+            break;
+        }
+        default:
+            break;
+    }
+
+    return rv;
```
Comment (BorjaOuterelo) [inline src/cpp/topic/Topic.cpp]:
```suggestion
    return (0 == std::strcmp(generic_type_.getName(), new_attributes.getTopicDataType().data())) &&
                     (generic_type_.m_isGetKeyDefined == (new_attributes.getTopicKind() ==
                                                          fastrtps::rtps::TopicKind_t::WITH_KEY));;
```

File: test/blackbox/tree/TreeTests.cpp [hunk 5]
```diff
@@ -267,6 +267,676 @@ TEST_F(TreeTests, REFTree)
...
+    response = client->create(creation_mode, participant_id, object_variant);
+    ASSERT_EQ(dds::xrce::STATUS_OK, response.status());
+
+    response = client->delete_object(participant_id);
+    creation_mode.reuse(true);
+    creation_mode.replace(true);
+    response = client->create(creation_mode, participant_id, object_variant);
+    ASSERT_EQ(dds::xrce::STATUS_OK, response.status());
+
+    creation_mode.reuse(false);
```
Comment (BorjaOuterelo) [inline test/blackbox/tree/TreeTests.cpp]: From here maybe split in multiple tests.

File: src/cpp/client/ProxyClient.cpp [hunk 6]
```diff
@@ -372,38 +314,70 @@ bool ProxyClient::create_datareader(const dds::xrce::ObjectId& object_id,
...
-                default:
-                    break;
+                rv = objects_.insert(std::make_pair(object_id, std::move(datareader))).second;
             }
         }
     }
     return rv;
 }
 
+bool ProxyClient::object_matched(std::shared_ptr<XRCEObject>& old_object,
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply src/cpp/client/ProxyClient.cpp]: Thanks for your piece of advice, it was implemented.

File: src/cpp/datareader/DataReader.cpp (hunk 1 above)
Comment (julianbermudez) [reply src/cpp/datareader/DataReader.cpp]: Done!

File: src/cpp/datawriter/DataWriter.cpp (hunk 2 above)
Comment (julianbermudez) [reply src/cpp/datawriter/DataWriter.cpp]: Done!

File: src/cpp/participant/Participant.cpp (hunk 3 above)
Comment (julianbermudez) [reply src/cpp/participant/Participant.cpp]: Done!

File: src/cpp/topic/Topic.cpp (hunk 4 above)
Comment (julianbermudez) [reply src/cpp/topic/Topic.cpp]: Done!

File: test/blackbox/tree/TreeTests.cpp (hunk 5 above)
Comment (julianbermudez) [reply test/blackbox/tree/TreeTests.cpp]: It will be simplify for the next release.

**Pull request merged**